### PR TITLE
Extract the cycle detector.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: src/compiler/cycle_detector.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+template<typename T>
+class CycleDetector {
+ public:
+
+  int in_progress_size() const {
+    return static_cast<int>(_in_progress.size());
+  }
+
+  void start(const T& entry) {
+     ASSERT(_in_progress_map.find(entry) == _in_progress_map.end());
```
Comment (kasperl) [inline src/compiler/cycle_detector.h:36]: Weird indentation?

File: src/compiler/cycle_detector.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+#include "map.h"
+#include "set.h"
+
+namespace toit {
+namespace compiler {
+
+template<typename T>
+class CycleDetector {
+ public:
+
```
Comment (kasperl) [inline src/compiler/cycle_detector.h:30]: Drop newline?

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: src/compiler/cycle_detector.h (hunk 2 above)
Comment (floitsch) [reply src/compiler/cycle_detector.h:30]: done.

File: src/compiler/cycle_detector.h (hunk 1 above)
Comment (floitsch) [reply src/compiler/cycle_detector.h]: done. (and removed the empty line 39).

**Pull request merged**